README.md
# php-rssCloud

An rssCloud server implementation in PHP

This has been inspired heavily by Andrew Shell's [server in node.js](https://github.com/rsscloud/rsscloud-server).

A key difference here is that subscriptions are **not** removed after 25 hours so do not need to be refreshed. See POST /cancel/ below.  
I am considering situations where subscriptions might be auto-removed after X notification failures (a failures column already exists in the subs table but is not currently used) but have yet to make any final decisions.

## Prerequisites

- a MySQL database
- an account to read/write from/to that database
    - it is preferable to have separate accounts for read & write
    
## Setup

- copy all files to the required location
- add database and account details to /config/config.php

```
define('DB_SERVER', '');
define('DB_NAME', '');
define('DB_USERNAME', '');
define('DB_PASSWORD', '');
define('DB_USERNAMESEL', '');
define('DB_PASSWORDSEL', '');
```

(DB_USERNAME & DB_PASSWORD for writing, DB_USERNAMESEL & DB_PASSWORDSEL for read only)

- create the required tables:
    - use the SQL statements in tables.txt
    
## Usage

### POST /pleaseNotify/

Posting to /pleaseNotify/ tells the server that you wish to receive updates from the specified feed.

The required parameters are:

1.  **domain** - the domain of your notification endpoint. This can be just the host portion (e.g. mysite.com) or you can also include the URL scheme (http/https).
  
3.  **port** - the port your endpoint uses. 80 & 8080 are treated as http, 443 is treated as https. Otherwise, the default is http but adding the scheme to the domain will override the port.
  
5.  **path** - the path of your notification endpoint relative to the domain above (e.g. /notify/).
  
7.  **url** - the address of the feed you wish to receive notifications for.

The rssCloud spec also includes 'registerProcedure' and 'protocol' parameters. These are both included but hidden. registerProcedure is not used and this implementation only supports http-post.

When you POST to **/pleaseNotify/** the server checks the link validity then performs a GET request to the notification endpoint represented by http://: with two query string parameters: url and challenge. To accept the subscription the endpoint needs to return an HTTP 200 status code and have the challenge value as the response body.

/pleaseNotify/ will return an XML request body with success="true" or "false", for example:

    <?xml version="1.0" ?>
    <notifyResult success="true"/>

### POST /ping/

Posting to /ping/ tells the server that the specified feed has been updated.

The required parameters are:

1.  **url** - the full address of the feed

When you POST to **/ping/** the server checks if the feed has actually changed since the last ping. If so, it will cycle through the subscribers to that feed and POST to the respective endpoints with the parameter url.

/ping/ will return an XML request body with success="true" or "false" depending on whether the feed has changed since the last ping.

If the server fails to notify any given subscriber it will increment the 'failures' column in the rssc_sub table. If it reaches 5 failures the subscription will be removed.

### POST /cancel/

Posting to /cancel/ tells the server that you wish to stop receiving updates from the specified feed.

The required parameters are:

1.  **domain** - the domain of your notification endpoint.
  
3.  **path** - the path of your notification endpoint relative to the domain.
  
5.  **url** - the full address of the feed you wish to unsubscribe from.

When you POST to **/cancel/** the server checks for a matching subscriber/feed pair and deletes it if found.

/cancel/ will return an XML request body with success="true" or "false".
  
  
**IMPORTANT**: please ensure you include the trailing slashes for the above endpoints.

## Forms

The server has web based forms that you can use instead of taking code-based actions.

**/pingForm/**, **/pleaseNotifyForm/**, and **/cancelForm/** each correspond to their respective endpoints above.

## Log

Recent events can be seen at **/viewLog/**.

After manually filling out each of the forms you will be automatically routed to the log to see whether the action succeeded or failed.

## To do

- clearing old events from the log, I need to determine a suitable retention period


changelog.md
# php-rssCloud changelog

**2nd March 2023**

Adjust url parameter check. More to come.

Updated files:

- .pleaseNotify/index.php

**27th Feb 2023**

Added logging for auto-removal of subs due to 5 failures.

Made a new icon and added to each page.

Updated files:

index.html
style.css
/cancelForm/index.html
/docs/index.html
/images/cloud.png
/ping/index.php
/pingForm/index.html
/pleaseNotifyForm/index.html
/viewLog/index.php

**26th Feb 2023**

If the server fails to notify any given subscriber it will increment the 'failures' column in the rssc_sub table for that subscription. If it reaches 5 failures the subscription will be removed automatically.

Updated files:

/ping/index.php


config/config.php
<?php

if(!defined('APP_RAN')) { die(); }

define('DB_SERVER', '');
define('DB_NAME', '');
define('DB_USERNAME', '');
define('DB_PASSWORD', '');
define('DB_USERNAMESEL', '');
define('DB_PASSWORDSEL', '');

class DBR {
	private $handle;
	
	public function __construct() {
		try {
			$this->handle = new mysqli(DB_SERVER, DB_USERNAMESEL, DB_PASSWORDSEL, DB_NAME);
		} catch (exception $e) {
		}
	}
	
	public function read($query) {
	    if ($result = $this->handle->prepare($query)) {
			$result = $this->handle->query($query)->fetch_all(MYSQLI_ASSOC);
		} else {
			$result = [];
    	}
    	return $result;
	}
	
	function __destruct() {
		try {
		    $this->handle->close();
		    $this->handle = null;
		} catch (exception $e) {
		}
	}
}

class DBW {
	private $handle;
	
	public function __construct() {
		try {
			$this->handle = new mysqli(DB_SERVER, DB_USERNAME, DB_PASSWORD, DB_NAME);
		} catch (exception $e) {
		}
	}
	
	public function write($query) {
		$this->handle->query($query);
		$insert_id = $this->handle->insert_id;
		return $insert_id;
	}
	
	function __destruct() {
		try {
		    $this->handle->close();
		    $this->handle = null;
	   	} catch (exception $e) {
		}
	}
}

function readdb($query) {
	$dbr = new DBR();
	return $dbr->read("$query");
}

?>


pleaseNotify/index.php
<?php

define('APP_RAN', '');

require_once('../config/config.php');

if ($_SERVER["REQUEST_METHOD"] === "POST") {

	try {
		$domain = $_POST["domain"];
		$port = $_POST["port"];
		$path = $_POST["path"];
		$registerProcedure = $_POST["registerProcedure"];
		$protocol = $_POST["protocol"];
		if (isset($_POST['url'])) {
		    $url = $_POST['url'];
		} else if (isset($_POST['url1'])) {
		    $url = $_POST['url1'];
		}
	} catch (exception $e) {
	}
	
	switch ($port) {
		case "80":
			$scheme = 'http';
			break;
		case "8080":
			$scheme = 'http';
			break;
		case "443":
			$scheme = 'https';
			break;
		default:
			$scheme = 'http';
	}
	
	$ds = parse_url($domain, PHP_URL_SCHEME);
	$ds != '' ? $scheme = $ds : null ;
	
	try {
		$headers = get_headers($url);
		if (substr($headers[0], 9, 3) == '200') {
			// feed http code 200 OK
			$fail = false;
		} else {
			// can't find feed
			$fail = true;
		}
	} catch (exception $e) {
		echo $e;
		$fail = true;
	}
	
	try {
		$endpoint = $scheme.'://'.$domain.$path;
		$headers = get_headers($endpoint);
		
		if (substr($headers[0], 9, 3) == '200') {
			// endpoint http code 200 OK
			$challenge = bin2hex(random_bytes(16));
			$qry_str = '?url='.$url.'&challenge='.$challenge;
			
			$options = array(
				CURLOPT_RETURNTRANSFER => TRUE,
				CURLOPT_FOLLOWLOCATION => TRUE,
				CURLOPT_PORT => $port,
				
			);
			
			$ch = curl_init();
			curl_setopt($ch, CURLOPT_URL,$endpoint.$qry_str);
			curl_setopt_array($ch, $options);
			$result = curl_exec($ch);
	        if ($result === false || $result != $challenge) {
	            $fail = true;
	            $dbw = new DBW();
	            $dbw->write("INSERT INTO rssc_log (endpoint, feed, action, status) VALUES ('$endpoint', '$url', 1, 0)");
	        }
	        curl_close($ch);
	        $fail = false;
		} else {
			// can't find endpoint
			$fail = true;
			$msg = "can't find endpoint: ".$endpoint;
		}
	} catch (exception $e) {
		$fail = true;
		$msg = "curl failed";
	}
	
	if ($fail === false) {
		//check for duplicate
		echo '<?xml version="1.0" ?>';
		echo '<notifyResult success="true"/>';
		$dbr = readdb("SELECT * FROM rssc_sub WHERE endpoint='$endpoint' AND feed='$url'");
		if (count($dbr) > 0) {
			$time = time();
			$dbw = new DBW();
			$dbw->write("UPDATE rssc_sub SET time='$time' WHERE endpoint='$endpoint' AND feed='$url'");
	        $dbw = new DBW();
	        $dbw->write("INSERT INTO rssc_log (time, endpoint, feed, action, status) VALUES ('$time', '$endpoint', '$url', 1, 3)");
		} else {
			$time = time();
			$dbw = new DBW();
			$dbw->write("INSERT INTO rssc_sub (time, endpoint, port, feed, failures) VALUES ('$time', '$endpoint', '$port', '$url', 0)");
			$dbw = new DBW();
	        $dbw->write("INSERT INTO rssc_log (time, endpoint, feed, action, status) VALUES ('$time', '$endpoint', '$url', 1, 1)");
		}
	} else {
		echo '<?xml version="1.0" ?>';
		echo '<notifyResult success="false" msg="'.$msg.'"/>';
		$time = time();
        $dbw = new DBW();
        $dbw->write("INSERT INTO rssc_log (time, endpoint, feed, action, status) VALUES ('$time', '$endpoint', '$url', 1, 0)");
	}
}

//redirect to log if manual sub from form

if (isset($_POST["notifyForm"])) {
?>
<script>
window.location.href = '../viewLog/';
</script>
<?php } ?>